### ISSUE TITLE
dvc exp run -S: -S not overriding yaml variables

### ISSUE THREAD

**Issue by f-fuchs** [treeverse/dvc#10984]
# Bug Report

## Description

-S is not correctly overriding yaml variables. Maybe the substitution is happening at the wrong step?

### Reproduce

params.yaml
```
training:
  epochs: 100
  gradient_max_norm: 1
  patience: 25
  lr: &lr 0.001
  w1: &w1 1.0
  w2: &w2 0.83
  w3: &w3 2500
  loss:
    aggregator:
      type: null
      params:
        weights: [*w1, *w2, *w3] 
```

Command with override for training.w3, correctly overrides training.w3 but training.loss.aggregator.params.weights where the variable w3 is used was not changed.
```
dvc exp run -S training.w3=5200.0                                                                                                                                                                                                                                                                                                                                                     
Reproducing experiment 'winey-jive'                                                                                                                                                                                                                                                                                                                                                                                                                             
Building workspace index                                                                                                                                                                                                                                                                                                                                                                                                              |5.00 [00:00,  221entry/s]
Comparing indexes                                                                                                                                                                                                                                                                                                                                                                                                                     |41.0 [00:00,  962entry/s]
Applying changes                                                                                                                                                                                                                                                                                                                                                                                                                      |23.0 [00:00,   126file/s]
Running stage 'train':                                                                                                                                                                                                                                                                                                                                                                                                                                          
> python train.py                                                                                                                                                                                                                                                                                                                                                                                                                                               
/home/fuchsfa/wavelet-extraction-model/.venv/lib/python3.12/site-packages/torch/cuda/__init__.py:65: FutureWarning: The pynvml package is deprecated. Please install nvidia-ml-py instead. If you did not install pynvml directly, please report this to the maintainers of the package that installed pynvml for you.                                                                                                                                          
  import pynvml  # type: ignore[import]                                                                                                                                                                                                                                                                                                                                                                                                                         
2026-01-30 10:46:06 [info     ] train.parse_config             config={'data': {'dataset': {'type': 'NumpyDataset', 'params': {'reflectivities': '/home/scratch/fuchs/welltie/300000_0.002_149_100_55__30_01_2026_09_44_31/reflectivities.npy', 'seismics': '/home/scratch/fuchs/welltie/300000_0.002_149_100_55__30_01_2026_09_44_31/seismics.npy', 'wavelets': '/home/scratch/fuchs/welltie/300000_0.002_149_100_55__30_01_2026_09_44_31/wavelets.npy'}}, 'spl
it': {'random': False, 'train_percentage': 0.8, 'val_percentage': 0.2}, 'data_loader': {'batch_size': 512, 'num_workers': 4, 'persistent_workers': True, 'pin_memory': True, 'shuffle': True, 'multiprocessing_context': 'spawn'}}, 'model': {'checkpoint': None, 'device': 'cuda_auto', 'type': 'VariationalNetwork', 'params': {'wavelet_size': 149, 'dropout_rate': 0.05, 'kernel_size': 3, 'downsampling_factor': 2, 'latent_dimensions': 128}}, 'determinis
tic': {'random_seed': 0, 'use_deterministic_algorithms': True, 'dataloaders': True, 'cublas': True}, 'training': {'epochs': 100, 'gradient_max_norm': 1, 'patience': 25, 'lr': 0.001, 'w1': 1.0, 'w2': 0.83, 'w3': 5200.0, 'loss': {'aggregator': {'type': None, 'params': {'weights': [1.0, 0.83, 2500]}}, 'losses': [{'L1Loss': {'reduction': 'sum'}}, {'KLDivergenceLoss': None}, {'FFTLoss': {'mask': 50}}]}, 'lr_scheduler': {'name': 'OneCycleLR', 'max_lr
': 0.001, 'total_steps': 'PLACE_HOLDER_CALCULATED_TOTAL_IT'}, 'optimizer': {'name': 'Adam', 'weight_decay': 0.003}}, 'plotting': {'wavelet_indices': [0, 1000, 2000, 3000, 4000, 5000, 10000, 20000], 'interval': 5, 'autotie_datasets': ['./autotie_datasets/sverdrup_16_2_21', './autotie_datasets/sverdrup_16_3_06', './autotie_datasets/torosa'], 'expected_sampling_time': 0.002}}  
```

### Expected

Places where the variables are substituted should change as well.

### Environment information

<!--
This is required to ensure that we can reproduce the bug.
-->

**Output of `dvc doctor`:**

```console
dvc doctor
DVC version: 3.66.1 (pip)
-------------------------
Platform: Python 3.12.7 on Linux-5.15.0-164-generic-x86_64-with-glibc2.41
Subprojects:
        dvc_data = 3.18.2
        dvc_objects = 5.2.0
        dvc_render = 1.0.2
        dvc_task = 0.40.2
        scmrepo = 3.6.1
Supports:
        http (aiohttp = 3.13.3, aiohttp-retry = 2.9.1),
        https (aiohttp = 3.13.3, aiohttp-retry = 2.9.1)
Config:
        Global: /home/fuchsfa/.config/dvc
        System: /etc/xdg/dvc
Cache types: hardlink, symlink
Cache directory: beegfs on beegfs_nodev
Caches: local
Remotes: None
Workspace directory: beegfs on beegfs_nodev
Repo: dvc, git
Repo.site_cache_dir: /var/tmp/dvc/repo/1733613c0f2eecf1ddb981adaf4d9239
```

**Comment by skshetry** [comment]
I think this behavior is working as expected. When YAML files are parsed, anchors and aliases are resolved immediately into their actual values. By the time DVC processes the parameters, there's no longer any connection between `training.w3` and the values in `training.loss.aggregator.params.weights`, they've become independent values.

Preserving anchors/aliases are done by roundtrip parser of `ruamel.yaml`.
Whether or not this should be supported by `ruamel.yaml` roundtrip parsing, you'd have to ask the maintainer of that library. However, I don't intend to add support for this use case in DVC.

**Comment by f-fuchs** [comment]
Is there another way then do this in dvc? can I update only elements of the weights list with -S?
